qiskit/transpiler/passes/optimization/drop_negligible.py

```diff
@@ -0,0 +1,70 @@
+# This code is part of Qiskit.
+#
+# (C) Copyright IBM 2024.
+#
+# This code is licensed under the Apache License, Version 2.0. You may
+# obtain a copy of this license in the LICENSE.txt file in the root directory
+# of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+#
+# Any modifications or derivative works of this code must retain this
+# copyright notice, and modified files need to carry a notice indicating
+# that they have been altered from the originals.
+
+"""Transpiler pass to drop gates with negligible effects."""
+
+from __future__ import annotations
+
+import numpy as np
+from qiskit.circuit.library import (
+    CPhaseGate,
+    PhaseGate,
+    RXGate,
+    RXXGate,
+    RYGate,
+    RYYGate,
+    RZGate,
+    RZZGate,
+    XXMinusYYGate,
+    XXPlusYYGate,
+)
+from qiskit.dagcircuit import DAGCircuit
+from qiskit.transpiler.basepasses import TransformationPass
+
+# List of Gate classes with the property that if the gate's parameters are all
+# (close to) zero then the gate has (close to) no effect.
+DROP_NEGLIGIBLE_GATE_CLASSES = (
+    CPhaseGate,
+    PhaseGate,
+    RXGate,
+    RYGate,
+    RZGate,
+    RXXGate,
+    RYYGate,
+    RZZGate,
+    XXPlusYYGate,
+    XXMinusYYGate,
+)
+
+
+class DropNegligible(TransformationPass):
+    """Drop gates with negligible effects."""
+
+    def __init__(self, atol: float = 1e-8) -> None:
+        """Initialize the transpiler pass.
+
+        Args:
+            atol: Absolute numerical tolerance for determining whether a gate's effect
+                is negligible.
+        """
+        self.atol = atol
+        super().__init__()
+
+    def run(self, dag: DAGCircuit) -> DAGCircuit:
+        for node in dag.op_nodes():
+            if not isinstance(node.op, DROP_NEGLIGIBLE_GATE_CLASSES):
+                continue
```

Comment:
Since you're using a fixed set of gates in this PR one thing you should do here is before looping is check if the names of the gate classes intersect with `dag.count_ops()`. That'll be much faster in the negative case.

Reply:
Doesn't `dag.count_ops` also need to iterate through the nodes of the dag?

Reply:
`DAGCircuit` maintains a dictionary of counts that it updates whenever an op is added or removed from the circuit. This is explicitly to support cases like this to do constant time checks for the instructions presence in the circuit. The only exceptions are if you set `recurse=True` (which is the default) and there are any control flow ops in the circuit.

Reply:
`dag.count_ops` stores strings, but I'm using the gate class itself. Any advice? Should I also just use strings? That doesn't seem robust because any gate can have any arbitrary name. But I guess there's many places in Qiskit which assume to know the type of a gate based on its name.